plugins/PostProcessingPlugin/scripts/SearchAndReplace.py

```diff
@@ -23,37 +23,170 @@ def getSettingDataString(self):
+                    "type": "bool",
+                    "default_value": true,
+                    "enabled": "not enable_layer_search"
+                },
+                "ignore_end":
+                {
+                    "label": "Ignore Ending G-code:",
+                    "description": "When enabled the Ending Gcode is unaffected.",
+                    "type": "bool",
+                    "default_value": true,
+                    "enabled": "not enable_layer_search"
                 }
             }
         }"""
 
     def execute(self, data):
+        global_stack = Application.getInstance().getGlobalContainerStack()
+        extruder = global_stack.extruderList
+        retract_enabled = bool(extruder[0].getProperty("retraction_enable", "value"))
+        # If retractions are enabled then the CuraEngine inserts a single data item for the retraction at the end of the last layer
+        # 'top_layer' accounts for that
+        if retract_enabled:
+            top_layer = 2
+        else:
+            top_layer = 1
         search_string = self.getSettingValueByKey("search")
-        if not self.getSettingValueByKey("is_regex"):
-            search_string = re.escape(search_string) #Need to search for the actual string, not as a regex.
-        search_regex = re.compile(search_string)
-
         replace_string = self.getSettingValueByKey("replace")
+        is_regex = self.getSettingValueByKey("is_regex")
+        enable_layer_search = self.getSettingValueByKey("enable_layer_search")
+        start_layer = self.getSettingValueByKey("search_start")
+        end_layer = self.getSettingValueByKey("search_end")
+        ignore_start = self.getSettingValueByKey("ignore_start")
+        ignore_end = self.getSettingValueByKey("ignore_end")
+        if enable_layer_search:
+            ignore_start = True
+            ignore_end = True
+        first_instance_only = bool(self.getSettingValueByKey("first_instance_only"))
 
-        for layer_number, layer in enumerate(data):
-            data[layer_number] = re.sub(search_regex, replace_string, layer) #Replace all.
+        #Find the raft and layer:0 indexes
+        raft_start_index = 0
+        layer_0_index = 0
+        start_index = 1
+        end_index = len(data)
+        try:
+            for l_num in range(2,12,1):
+                layer = data[l_num]
+                if ";LAYER:-" in layer and raft_start_index == 0:
+                    raft_start_index = l_num
+                if ";LAYER:0" in layer:
+                    layer_0_index = l_num
+                    break
+            if raft_start_index == 0:
+                raft_start_index = layer_0_index
+                raft_layers = 0
+            elif raft_start_index < layer_0_index:
+                raft_layers = layer_0_index - raft_start_index
+            else:
+                raft_layers = 0
+        except:
```

Comment:
If this `except` is supposed to catch a specific case, please define the type of caught exception (`KeyError` probably ?) and add a small comment